src/zon.zig

```diff
@@ -0,0 +1,831 @@
+        .tuple_type => self.lowerTuple(node, res_ty),
+        .struct_type => self.lowerStruct(node, res_ty),
+        else => unreachable,
+    };
+}
+
+fn lowerTuple(self: LowerZon, node: Zoir.Node.Index, res_ty: Type) !InternPool.Index {
+    const ip = &self.sema.pt.zcu.intern_pool;
+
+    const tuple_info = ip.indexToKey(res_ty.toIntern()).tuple_type;
+
+    const elem_nodes: Zoir.Node.Index.Range = switch (node.get(self.file.zoir.?)) {
+        .array_literal => |nodes| nodes,
+        .empty_literal => .{ .start = node, .len = 0 },
+        else => return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "expected type '{}'",
+            .{res_ty.fmt(self.sema.pt)},
+        ),
+    };
+
+    const field_defaults = tuple_info.values.get(ip);
+    const field_types = tuple_info.types.get(ip);
+    const elems = try self.sema.arena.alloc(InternPool.Index, field_types.len);
+    for (elems) |*v| v.* = .none;
+
+    for (0..elem_nodes.len) |i| {
+        if (i >= elems.len) {
+            const elem_node = elem_nodes.at(@intCast(i)).getAstNode(self.file.zoir.?);
+            return self.fail(
+                .{ .node_abs = elem_node },
+                "index {} outside tuple of length {}",
+                .{
+                    elems.len,
+                    elem_nodes.at(@intCast(i)).getAstNode(self.file.zoir.?),
+                },
+            );
+        }
+        elems[i] = try self.lowerExpr(elem_nodes.at(@intCast(i)), Type.fromInterned(field_types[i]));
+
+        if (field_defaults[i] != .none and elems[i] != field_defaults[i]) {
+            const elem_node = elem_nodes.at(@intCast(i)).getAstNode(self.file.zoir.?);
+            return self.fail(
+                .{ .node_abs = elem_node },
+                "value stored in comptime field does not match the default value of the field",
+                .{},
+            );
+        }
+    }
+
+    for (0..elems.len) |i| {
+        if (elems[i] == .none and i < field_defaults.len) {
+            elems[i] = field_defaults[i];
+        }
+    }
+
+    for (elems, 0..) |val, i| {
+        if (val == .none) {
+            return self.fail(
+                .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                "missing tuple field with index {}",
+                .{i},
+            );
+        }
+    }
+
+    return self.sema.pt.intern(.{ .aggregate = .{
+        .ty = res_ty.toIntern(),
+        .storage = .{ .elems = elems },
+    } });
+}
+
+fn lowerStruct(self: LowerZon, node: Zoir.Node.Index, res_ty: Type) !InternPool.Index {
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const gpa = self.sema.gpa;
+
+    try res_ty.resolveFields(self.sema.pt);
+    try res_ty.resolveStructFieldInits(self.sema.pt);
+    const struct_info = self.sema.pt.zcu.typeToStruct(res_ty).?;
+
+    const fields: std.meta.fieldInfo(Zoir.Node, .struct_literal).type = switch (node.get(self.file.zoir.?)) {
```

Comment:
```suggestion
    const fields: @FieldType(Zoir.Node, "struct_literal") = switch (node.get(self.file.zoir.?)) {
```
To be fair, that builtin didn't exist when you started this PR :P